scripts/build.py

```diff
@@ -174,47 +184,113 @@ def _ensure_files_exist(filepaths):
         filepaths: list(str). Paths to files that we want to ensure exist.
 
     Raises:
-        Exception: Raised if one or more of the files does not exist.
+        OSError: One or more of the files does not exist.
     """
     for filepath in filepaths:
         if not os.path.isfile(filepath):
-            raise Exception('File %s does not exist.' % filepath)
+            raise OSError('File %s does not exist.' % filepath)
+
+
+def get_file_count(directory_path):
+    """Count total number of file directory, subtracting ignored files.
+
+    Args:
+        directory_path: str. Directory to be walked.
 
+    Returns:
+        int. Total number of files minus ignored files.
+    """
+    total_file_count = 0
+    for _, _, files in os.walk(directory_path):
+        total_file_count += len(files)
+        for filename in files:
+            # Ignore files with certain extensions.
+            if any(filename.endswith(p)
+                   for p in FILE_EXTENSIONS_TO_IGNORE):
+                total_file_count -= 1
+    if directory_path == ASSETS_DEV_DIR:
+        # Add 1 to account for the added hashes.js.
+        total_file_count += 1
+    return total_file_count
 
-def process_html(source_path, target_path, file_hashes):
+
+def _compare_file_count(source_path, target_path):
+    """Ensure that two dir's file counts match.
+
+    Args:
+       source_path: str. Source of staging files.
+       target_path: str. Final filepath or built file's directory.
+
+    Raises:
+        ValueError: Raised if both dir do not have the same file count.
+    """
+    source_dir_file_count = get_file_count(source_path)
+    target_dir_file_count = get_file_count(target_path)
+    if source_dir_file_count != target_dir_file_count:
+        print 'Comparing %s vs %s' % (source_path, target_path)
+        raise ValueError(
+            '%s files in source dir != %s files in target dir.' % (
+                source_dir_file_count, target_dir_file_count))
+
+
+def _match_filename_with_hashes(filename, file_hashes):
+    """Ensure that filepath is hashed correctly.
```

Comment:
Let's be a bit clearer about what "hashed correctly" means -- I don't entirely understand what's being checked here. Are we trying to check the filename, or the file hashes, or both?

It would help to clarify the function name and docstring. If it's difficult to do that clearly, then it's possible that the function is checking too many things and we should split it up a bit.

Reply:
Clarified `_match_filename_with_hashes()` docstring.
Please lmk if it is still unclear from your perspective!